src/recap.c

```diff
@@ -0,0 +1,490 @@
+
+  if ((read_size = read_pcap(pctx, len)) < 0) {
+    log_error("read_pcap fail");
+    return -1;
+  }
+
+  if (pctx->data_size >= pctx->pkt_header.caplen) {
+    log_trace("Received pkt data");
+
+    if (save_packet(pctx) < 0) {
+      log_error("process_packet fail");
+      return -1;
+    }
+
+    pctx->data_size = 0;
+    pctx->state = PCAP_STATE_READ_PKT_HEADER;
+  } else if (read_size == 0) {
+    log_trace("No data received");
+    pctx->state = PCAP_STATE_FIN;
+  }
+
+  return 0;
+}
+
+int process_pcap_stream_state(struct pcap_stream_context *pctx) {
+  log_trace("Processing pcap file stream %zu bytes", pctx->total_size);
+
+  switch (pctx->state) {
+    case PCAP_STATE_INIT:
+      if ((pctx->pcap_data = os_malloc(sizeof(char))) == NULL) {
+        log_errno("os_malloc");
+        return -1;
+      }
+      pctx->npackets = 0;
+      pctx->total_size = 0;
+      pctx->data_size = 0;
+      pctx->state = PCAP_STATE_READ_PCAP_HEADER;
+      return 1;
+    case PCAP_STATE_READ_PCAP_HEADER:
+      if (process_pcap_header_state(pctx) < 0) {
+        log_error("process_pcap_header_state fail");
+        return -1;
+      }
+      return 1;
+    case PCAP_STATE_READ_PKT_HEADER:
+      if (process_pkt_header_state(pctx) < 0) {
+        log_error("process_pkt_header_state fail");
+        return -1;
+      }
+      return 1;
+    case PCAP_STATE_READ_PACKET:
+      if (process_pkt_read_state(pctx) < 0) {
+        log_error("process_pkt_read_state fail");
+        return -1;
+      }
+      return 1;
+    case PCAP_STATE_FIN:
+      return 0;
+    default:
+      log_trace("Unknown state");
+      return -1;
+  }
+}
+
+int main(int argc, char *argv[]) {
+  uint8_t verbosity = 0;
+  uint8_t level = 0;
+  char *pcap_path = NULL, *db_path = NULL;
+  struct pcap_stream_context pctx = {.db = NULL,
+                                     .pcap_fd = NULL,
+                                     .ifname = NULL,
+                                     .state = PCAP_STATE_INIT,
+                                     .total_size = 0,
+                                     .npackets = 0};
+
+  process_app_options(argc, argv, &verbosity, &pcap_path, &db_path,
+                      &pctx.ifname);
+
+  if (verbosity > MAX_LOG_LEVELS) {
+    level = 0;
+  } else if (!verbosity) {
+    level = MAX_LOG_LEVELS - 1;
+  } else {
+    level = MAX_LOG_LEVELS - verbosity;
+  }
+
+  if (optind <= 1) {
+    show_app_help(argv[0]);
+  }
+
+  if (pctx.ifname == NULL) {
+    pctx.ifname = os_strdup(IFNAME_DEFAULT);
+  }
+
+  /* Set the log level */
+  log_set_level(level);
+
+  int ret = sqlite3_open(db_path, &pctx.db);
+
+  fprintf(stdout, "Openning db at %s\n", db_path);
+
+  if (ret != SQLITE_OK) {
+    fprintf(stdout, "Cannot open database: %s", sqlite3_errmsg(pctx.db));
+    if (pcap_path != NULL) {
+      os_free(pcap_path);
+    }
+    os_free(db_path);
+    os_free(pctx.ifname);
+    sqlite3_close(pctx.db);
+    return EXIT_FAILURE;
+  }
+
+  os_free(db_path);
```

Comment:
Might as well move `os_free(db_path)` out of the if loop.

To be honest, this function seems great to have a `goto exit_cleanup;`.

```suggestion
  os_free(db_path);

  if (ret != SQLITE_OK) {
    fprintf(stdout, "Cannot open database: %s", sqlite3_errmsg(pctx.db));
    if (pcap_path != NULL) {
      os_free(pcap_path);
    }
    os_free(pctx.ifname);
    sqlite3_close(pctx.db);
    return EXIT_FAILURE;
  }
```